server/src/main/java/io/crate/expression/scalar/ArrayPrependFunction.java

```diff
@@ -0,0 +1,93 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.crate.data.Input;
+import io.crate.metadata.FunctionType;
+import io.crate.metadata.Functions;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+import io.crate.types.TypeSignature;
+
+public class ArrayPrependFunction extends Scalar<List<Object>, Object> {
+
+    public static final String NAME = "array_prepend";
+
+    public static void register(Functions.Builder builder) {
+        builder.add(
+                Signature.builder(NAME, FunctionType.SCALAR)
+                        .argumentTypes(
+                            TypeSignature.parse("E"),
+                            TypeSignature.parse("array(E)")
+                            )
+                        .returnType(TypeSignature.parse("array(E)"))
+                        .typeVariableConstraints(typeVariable("E"))
+                        .features(Feature.DETERMINISTIC, Feature.NOTNULL)
+                        .build(),
+                ArrayPrependFunction::new
+        );
+    }
+
+    private final DataType<?> innerType;
+    private final boolean calledByOperator;
+
+    ArrayPrependFunction(Signature signature, BoundSignature boundSignature) {
+        this(signature, boundSignature, false);
+    }
+
+    ArrayPrependFunction(Signature signature, BoundSignature boundSignature, boolean calledByOperator) {
+        super(signature, boundSignature);
+        this.innerType = ((ArrayType<?>) boundSignature.returnType()).innerType();
+        this.calledByOperator = calledByOperator;
+    }
+
+    @Override
+    public final List<Object> evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input[] args) {
+        ArrayList<Object> resultList = new ArrayList<>();
+        Object valueToAdd = args[0].value();
+        @SuppressWarnings("unchecked")
+        List<Object> values = (List<Object>) args[1].value();
+
+        if (valueToAdd != null || calledByOperator == false) {
+            // null || array -> array (null is ignored)
+            resultList.add(innerType.sanitizeValue(valueToAdd));
+        }
```

Comment:
The opposite (not going into the if branch) is already tested by the existing null literal concat operator test at https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/expression/scalar/ConcatFunctionTest.java#L129. This matches now the new array_prepend` scalar.